Refactor `dm`

MalhotraPulak opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Plan:
- Refactor dm to use  RZ_API
- Refactor dmh to use RZ_API
- Make heap viewer in Cutter using the refactored commands

Currently:
- Add more tests for the commands
- Expose API to cutter

**Test plan**

CI is green

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

MalhotraPulak committed:
Remove non standard C

MalhotraPulak committed:
Print tcache and bin info in JSON

MalhotraPulak committed:
Add fd and bk pointers of non empty bins

MalhotraPulak committed:
Add more information about chunks

MalhotraPulak committed:
Run `dmhd` for different arenas

MalhotraPulak committed:
More info in `dmha` command and JSON version `dmhaj`

MalhotraPulak committed:
clang format

MalhotraPulak committed:
start making tree

MalhotraPulak committed:
dm./dmd/dmda/dmdw working

MalhotraPulak committed:
dm- working

MalhotraPulak committed:
dmh working

MalhotraPulak committed:
dm and dm= working

MalhotraPulak committed:
updated cmd_descs.c/h file

MalhotraPulak committed:
dmm working

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,24 +1443,145 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
 }
 
+// dm
+RZ_IPI RzCmdStatus rz_cmd_debug_list_maps_handler(RzCore *core, int argc, const char **input, RzOutputMode mode) {
```
I wonder if it's better to use `argv` instead of `input` as it's done in other places, to keep consistency. cc @ret2libc

MalhotraPulak replied:
yeah i will change that to `argv`

XVilka commented:
I suggest to port [that JSON PR](https://github.com/rizinorg/rizin/pull/1172) on top of this one once it's finished.

MalhotraPulak committed:
working on `dmi`

MalhotraPulak committed:
`dml` command working

MalhotraPulak committed:
change `input` -> `argv`

MalhotraPulak committed:
`dmL` working

MalhotraPulak committed:
`dmS` refactor

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,370 +1443,618 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+RZ_IPI RzCmdStatus rz_cmd_debug_list_maps_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	if (argc == 1) {
+		char str[5] = "";
+		if (argc == 1) {
```
It's already true here.

MalhotraPulak replied:
yes i will remove the check

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,370 +1443,618 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+	if (argc == 1) {
+		char str[5] = "";
+		if (argc == 1) {
+			strcpy(str, argv[0] + 2);
```
Are you sure `strcpy()` is necessary here? Maybe it will work without?

MalhotraPulak replied:
yeah it is not needed ill remove it

XVilka commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,176 @@ commands:
             args:
-            - name: addr
-              type: RZ_CMD_ARG_TYPE_RZNUM
+              - name: addr
```
Are you sure these formatting changes are necessary? I recommend to recheck using `yamllint`.

MalhotraPulak replied:
My ide autoformatted it,  i will run it  through `yamllint`

MalhotraPulak committed:
fixed formatting of yaml file

MalhotraPulak committed:
`dmS` moved from oldinput to regular

MalhotraPulak committed:
`dmq` command refactor

MalhotraPulak commented:
@XVilka I found `dmq`  and `dmq.` command in the codebase which were not mentioned in the help menu. I implemented that in the new shell. Can you tell what summary should I write for them?  
Also, in the help menu there are `dms` and `dms-` command mentioned but not implemented anywhere. Should I add them to new shell as commands that do nothing or leave them?

XVilka commented:
@PulakIIIT for
- `dms` - remove them
- `dmq` means `quiet`. Judging by the code it's the same as `dm` but with less verbose information. Just add ` RZ_OUTPUT_MODE_QUIET` in `dm` YAML and call it from here.

Moreover, I checked the code of `rz_debug_map_list()`. It is wrong nowadays because:
- Generally, when we have function ending in `_list()` it means it *returns* list of something, e.g. as `RzList *`. When the function *prints* something, we call it `_print()` instead.
- Generally, we prefer to avoid `RzCons` usage or printing outside of the `librz/core`.

Thus, I suggest to do the following:
- Add function `rz_debug_map_print()` in `librz/core/cdebug.c`
- Make function `rz_debug_map_list()` return `RzList *`
- Move all print helpers of old `rz_debug_map_list()` together with it into `librz/core/cdebug.c`
- Move `rz_debug_map_list_visual()` with helpers also into `librz/core/cdebug.c`
- Also I suggest to use `RzOutputMode` instead of `char` in that funciton like described in https://github.com/rizinorg/rizin/issues/489

MalhotraPulak committed:
`rz_debug_map_list` -> `rz_debug_map_print` refactor

MalhotraPulak commented:
@Xvilka I have moved the function `rz_debug_map_list` and related functions from `dmap.c` to `librz/core/cdebug.c` and renamed `rz_debug_map_list` to `rz_debug_map_print` since it mostly focuses on printing. I also refactored `rz_debug_map_print` to use `RzOutputMode`. Since I renamed that function at the moment there is no function  `rz_debug_map_list` which returns an `RzList *` as you mentioned in your first point. Is it necessary to have `rz_debug_map_list` in the codebase?

MalhotraPulak committed:
clang format

MalhotraPulak committed:
more clang format

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,270 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+}
+
+RZ_API void rz_debug_map_list_visual(RzDebug *dbg, ut64 addr, const char *input, int colors) {
+	if (dbg) {
```
Please reverse the condition, so there will be less nesting.

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,270 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+		int i;
+		for (i = 0; i < 2; i++) { // Iterate over dbg::maps and dbg::maps_user
+			RzList *maps = (i == 0) ? dbg->maps : dbg->maps_user;
+			if (maps) {
```
Same here.

XVilka commented:
>  Since I renamed that function at the moment there is no function rz_debug_map_list which returns an RzList * as you mentioned in your first point. Is it necessary to have rz_debug_map_list in the codebase?

Please add, you will need it for Cutter anyway.

MalhotraPulak committed:
add `rz_debug_list()`

XVilka commented on librz/core/cdebug.c:
```diff
@@ -885,24 +885,31 @@ static void print_debug_maps_ascii_art(RzDebug *dbg, RzList *maps, ut64 addr, in
 	}
 }
+
+RZ_API RzList *rz_debug_map_list(RzDebug *dbg, bool user_map) {
```
I recommend to move this function to `librz/debug/dmap.c` instead, it doesn't need to be in the core.
Please also add a short Doxygen description.

MalhotraPulak committed:
Add documentation and move `rz_debug_map_list()` ##doc

MalhotraPulak committed:
setting up `dmh` refactor

MalhotraPulak committed:
Add the `cmd_heap_glibc.yaml` file

MalhotraPulak committed:
`dmh` basic working and `dmhv` converted to `dmhl` ##refactor

MalhotraPulak committed:
`dmha` working ##refactor

MalhotraPulak committed:
`dmhm` working ##refactor

MalhotraPulak committed:
`dmhc` working ##refactor

MalhotraPulak committed:
`dmhj` and `dmh*` working ##refactor

MalhotraPulak committed:
`dmhg` working ##refactor

MalhotraPulak committed:
`dmhi` working ##refactor

MalhotraPulak committed:
fix a memory leak

MalhotraPulak committed:
Clang-format

MalhotraPulak committed:
`dmht` working ##refactor

MalhotraPulak committed:
`dmhb` working and more memory leaks fix ##refactor

MalhotraPulak committed:
update tests ##refactor

MalhotraPulak committed:
`dmhf`working ##refactor

MalhotraPulak committed:
`dmhd`working ##refactor

MalhotraPulak committed:
Remove old logic and check yamllint ##refactor

MalhotraPulak committed:
Fix a bug in `dmhl`

MalhotraPulak committed:
Add documentation for `get_heap_chunks_list`

MalhotraPulak committed:
Add documentation for `get_arenas_list` ##doc

MalhotraPulak committed:
Minor change in documentation ##doc

MalhotraPulak committed:
Invert ifs to decrease nested ifs ##refactor

MalhotraPulak committed:
Add status for free arena in `dmha` ##refactor

MalhotraPulak committed:
Enable windows and jemalloc parsing again, use `dmw` for windows heap commands and `dmx` for Jemalloc heap commands ##refactor

MalhotraPulak committed:
Separate printing and returning list for contents of a bin ##refactor

MalhotraPulak committed:
Separate init and printing of a heap chunk ##refactor

MalhotraPulak committed:
Added headers in rz_core.h ##refactor

MalhotraPulak committed:
Separate getting tcache info and printing it ##refactor

MalhotraPulak committed:
Fix bug in `dmht`

MalhotraPulak committed:
Add corrupted list check and more headers

XVilka commented:
Please also rebase on top of the latest dev to resolve the conflicts.

MalhotraPulak committed:
Fix error after rebase

MalhotraPulak committed:
Add more tests

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1494,544 +2061,336 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
+	free(main_arena);
+	return RZ_CMD_STATUS_OK;
+}
+#ifndef HEAP32
```
why this condition? Where is `rz_cmd_heap_chunks_print_handler` defined if this is not present?

MalhotraPulak replied:
`linux_heap_glibc.c` file has a recursive declaration of itself in line 10 
```C
#define HEAP32 1
#include "linux_heap_glibc.c"
#undef HEAP32
```
If I do not wrap these functions in a conditional statement then they would be declared twice.

ret2libc replied:
Move them to another file then... `cmd_linux_heap_glibc.c` or something like that. I think it's better.

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1494,544 +2061,336 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
-		return false;
+RZ_IPI RzCmdStatus rz_cmd_heap_chunks_graph_handler(RzCore *core, int argc, const char **argv) {
+	// RZ_OUTPUT_MODE_LONG_JSON mode workaround for graph
+	if (core->rasm->bits == 64) {
```
can't you just call `rz_cmd_heap_chunks_print_handler`?

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1494,544 +2061,336 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
+}
+#ifndef HEAP32
+RZ_IPI RzCmdStatus rz_cmd_heap_chunks_print_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	if (core->rasm->bits == 64) {
```
i'd suggest to define a small macro to avoid all these `if (core->rasm->bits == 64) ...._64 else ..._32`.

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1406,84 +1840,217 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```
I'd have this function with `GHT m_state` argument instead of argc/argv. That woulld make calling them "easier" for other parts of the code.

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -223,33 +221,33 @@ static const char *help_msg_dko[] = {
-	"TODO:", "", "map files in process memory. (dmf file @ [addr])",
-	NULL
-};
+//static const char *help_msg_dm[] = {
```
remember to just remove these if not used.

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,86 @@
+---
+name: cmd_heap_glibc
+commands:
+  - name: dmh
```
it would be great if you could use `type: RZ_CMD_DESC_TYPE_ARGV_STATE`. You can see `rz_analysis_function_blocks_list_handler`(and others) and read a bit the doc in `rz_cmd.h` on how that changes things.

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1406,84 +1840,217 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```
I think all these `dm` subcommands should have a nice check to see if debugging is enabled and provide a nice message to users if not.

MalhotraPulak replied:
@ret2libc Is this check good for debugging is enabled or you would suggest something else?
```C
	if (!core || !core->dbg || core->dbg->pid == -1) {
		rz_cons_printf("Rizin not in debugging mode. Run ood?\n");
		return RZ_CMD_STATUS_ERROR;
	}
```

ret2libc replied:
Seems like there is `rz_debug_is_dead` that can be used instead. cc @yossizap

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1436,408 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+}
+
+// dmp
+RZ_IPI RzCmdStatus rz_cmd_debug_dmp_handler(RzCore *core, int argc, const char **argv) {
```
i'd suggest a better name e.g. `rz_debug_memory_permission_handler`

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1436,408 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+	RzDebugMap *map;
+	ut64 addr = 0, size = 0;
+	int perms;
+	if (argc == 4) { // dmp <addr> <size> <perms>
```
these checks are not in sync with the optional/non-optional arguments defined in the .yaml. In particular, I see `dmp [<addr> [<size> <perms>]]`, but if i do `dmp` it says `Wrong number of arguments passed to `dmp`, see its help with `dmp?``.

MalhotraPulak replied:
`dmp` commands can be run in two ways:  
- `dmp <addr> <size> <perms>`
- `dmp <perms> `
How should the yaml file look like for input like this?

ret2libc replied:
let's change the way this command works IMHO. @XVilka opinions?

For example `dmp <perms> [<size>]` (again, I would not have `<addr>` in the arguments because that can be changed with `@ <addr>`). WDYT?

MalhotraPulak replied:
Yeah, this seems like a nice way to deal with the case.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,191 @@ commands:
+        summary: Allocate <size> bytes at <address> and promote to huge page
+        cname: cmd_debug_dmL
+        args:
+          - name: addr
```
I would avoid all `addr` arguments, in general, because I think `@ <addr>` modifier is already good enough. So I would make most commands just assume that they have to act on the current address (the user can always use `@ addr` to change it.

MalhotraPulak replied:
Should I add "use `@ <addr>` modifier" in the summary of the commands which use that or we just assume the user knows that?

ret2libc replied:
Let's just assume the user knows that. I think all commands should assume this behaviour and we should try to make it consistent everywhere, so users know they can use this pattern everywhere.

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,86 @@
+    cname: cmd_heap_chunk_print
+    summary: Get info about heap chunk at <addr>
+    args:
+      - name: addr
```
also here, just remove this argument and assume current offset.

MalhotraPulak committed:
Move some functions to new file `cmd_linux_heap_glibc.c`

MalhotraPulak committed:
Replaced repetitive code with a macro

MalhotraPulak committed:
Removed <addr> argument for some commands

MalhotraPulak committed:
Added debugging enabled check for all `dm` commands

MalhotraPulak committed:
Update test

MalhotraPulak committed:
Remove `<addr>` argument from `dm-`

MalhotraPulak committed:
Added `dma` command ##refactor

MalhotraPulak committed:
Fixed test

MalhotraPulak committed:
Renamed `dma` command to `dm+` ##refactor

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+      - name: dm+
+        cname: cmd_debug_allocate_maps
+        summary: >
+          Allocate <size> bytes at <offset>
```
Adjust the description, because `<offset>` does not make sense here.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+            summary: List memory map of current module
+            args: []
+      - name: dm-
+        summary: Deallocate memory map at <offset>
```
maybe "at current offset"

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+            optional: false
+      - name: dmp
+        summary: >
+          Change page at <offset> with <size>, protection <perms>
```
```suggestion
          Change page at current offset with <size>, protection <perms>
```

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+        args:
+          - name: perms
+            type: RZ_CMD_ARG_TYPE_STRING
+            optional: false
```
no need for this AFAIK. it's implicit.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+            type: RZ_CMD_ARG_TYPE_NUM
+            optional: true
+      - name: dmL
+        summary: Allocate <size> bytes at <offset> and promote to huge page
```
```suggestion
        summary: Allocate <size> bytes at current offset and promote to huge page
```

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+            type: RZ_CMD_ARG_TYPE_STRING
+            optional: false
+          - name: size
+            type: RZ_CMD_ARG_TYPE_NUM
```
```suggestion
            type: RZ_CMD_ARG_TYPE_RZNUM
```

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,185 @@ commands:
+        cname: cmd_debug_dmL
+        args:
+          - name: size
+            type: RZ_CMD_ARG_TYPE_NUM
```
```suggestion
            type: RZ_CMD_ARG_TYPE_RZNUM
```

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,84 @@
+    args: []
+  - name: dmhb
+    cname: cmd_heap_bins_list_print
+    summary: >
```
Maybe we should think about a shorter summary? You can also have the `description: ` field with a longer description.

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,84 @@
+      of main_arena's or a particular arena bins instance.
+      Use dmhbg command to get a graph of the linked list
+    type: RZ_CMD_DESC_TYPE_OLDINPUT
+    args:
```
if this is OLDINPUT type you don't need to specify the arguments.
Also, if you choose later to convert this to newshell, i would split `bin_num` and `malloc_state`.

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,84 @@
+        optional: true
+  - name: dmhc
+    cname: cmd_heap_chunk_print
+    summary: Get info about heap chunk at offset
```
```suggestion
    summary: Get info about heap chunk at current offset
```

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,84 @@
+      - RZ_OUTPUT_MODE_JSON
+    args:
+      - name: bin_type
+        type: RZ_CMD_ARG_TYPE_STRING
```
see CHOICE arg type. There are other examples of this type in .yaml files that you can check to see how to use it.

ret2libc commented on librz/core/cmd_descs/cmd_heap_glibc.yaml:
```diff
@@ -0,0 +1,84 @@
+    summary: >
+      Display all parsed fastbins of main_arena's or a particular
+      arena fastbinY instance
+    args:
```
OLDINPUT, thus args is not needed.

MalhotraPulak replied:
I used `args` to show them in help menu of the `dmh` command

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -770,6 +771,22 @@ RZ_API char *rz_core_sysenv_begin(RzCore *core, const char *cmd);
 
 RZ_API void rz_core_recover_vars(RzCore *core, RzAnalysisFunction *fcn, bool argonly);
+
+/* linux_heap_glibc.c */
```
The fact that the malloc implementations have to stay in RzCore seems a bit wrong, but definitely not something to consider in this PR.

However, do we really need these API exported  in the public API? Can we put them as RZ_IPI in core_private.h? If they have to stay in the public API they should be at least prefixed with `rz_core_`.

XVilka replied:
I agree that ideally they should be moved to `librz/debug`, except configuration handling and command handlers. But, indeed, this is out of the scope of this PR.

XVilka replied:
Also agree on keeping these APIs private.

XVilka replied:
@PulakIIIT don't forget about this one.

MalhotraPulak replied:
@XVilka @ret2libc I added them as public APIs as in the future I will need some of the malloc-related APIs when I implement the heap viewer in Cutter

XVilka replied:
Then I think you need better names for them. Something like `rz_debug_heap_` or `rz_heap_`. Currently it is quite inconsistent.

ret2libc replied:
Not that it really matters for the current state of things with regards to the API, but IMO even if you need them in the future it would be better to make them private for now and switch them to public only as you really need them (with proper unit tests if time/will permits).

Anyway, I agree for better names!

MalhotraPulak replied:
At the moment I have renamed the functions to be more consistent, in the following weeks I will be working on the heap viewer in Cutter so I will keep updating the Rizin API for heaps as I need and I will remove any API calls which are not required.

MalhotraPulak committed:
Implement suggested changes

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,273 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+					name, map->addr_end - map->addr + 1, map->addr);
+				free(name);
+			} break;
+			case RZ_OUTPUT_MODE_QUIET: // "dmq"
```
Please remove commented code.

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,273 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+			dbg->cb_printf("TODO:\n");
+			rz_list_foreach (maps, iter, map) {
+				if (addr >= map->addr && addr < map->addr_end) {
+					// print_debug_map_ascii_art (dbg, map);
```
Why this part is commented?

MalhotraPulak replied:
This is code for `dm=.` command and it was not fully implemented (therefore the `TODO` printf) even before the refactor. I can just remove this part since this code is not reachable after shifting this command to the new shell.

XVilka replied:
Yes, I think it's better to just remove that. If you think this is a feature is worth adding - just open an issue. But better to remove this dead code and TODO printing.

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,273 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+	}
+	int i;
+	for (i = 0; i < 2; i++) { // Iterate over dbg::maps and dbg::maps_user
+		//		RzList *maps = (i == 0) ? dbg->maps : dbg->maps_user;
```
Please remove this unnecessary commented code

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -270,12 +240,12 @@ static const char *help_msg_dmm[] = {
-	"dmp", " [perms]", "Change dbg.map permissions",
-	NULL
-};
+//static const char *help_msg_dmp[] = {
```
Just remove this

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1377,465 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+			/* TODO: do not spawn. use RzBin API */
+			if (sectname) {
+				char *sect = rz_str_escape(sectname);
+				res = rz_sys_cmd_strf("env RZ_BIN_PREFIX=\"%s\" rz-bin %s-B 0x%08" PFMT64x " -S \"%s\" | grep \"%s\"", name, mode, baddr, filesc, sect);
```
@thestr4ng3r could you please suggest a proper API as a substitute for this command?

wargio replied:
this is not ok.

wargio replied:
but we can change this in another PR

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -188,7 +188,14 @@ static void GH(update_arena_without_tc)(GH(RzHeap_MallocState) * cmain_arena, Ma
 
-static bool GH(update_main_arena)(RzCore *core, GHT m_arena, MallocState *main_arena) {
+/**
+ * @brief Store the MallocState struct of an arena with base address m_arena in main_arena
```
In other places we use `\brief` syntax instead of `@brief`. Please use it consistently. See https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#documentation

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -391,19 +398,26 @@ static void GH(print_arena_stats)(RzCore *core, GHT m_arena, MallocState *main_a
-	if (core->dbg->main_arena_resolved) {
-		return true;
-	}
+	//	if (core->dbg->main_arena_resolved) {
```
Why this was commented? Is this still necessary? If not, lets remove the commented code. cc @thestr4ng3r @ret2libc

MalhotraPulak replied:
During the refactoring, I tried to get the code with this statement uncommented but I was facing some bugs in which `m_arena` was not be initialized properly. Due to the nature of the refactor `GHT m_arena` is declared multiple times so we cannot use this hack to save performance anymore.

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -391,19 +398,26 @@ static void GH(print_arena_stats)(RzCore *core, GHT m_arena, MallocState *main_a
 
-static bool GH(rz_resolve_main_arena)(RzCore *core, GHT *m_arena) {
+/**
+ * @brief Store the base address of main arena at m_arena
```
Use `\` vs `@`

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1025,278 +1083,656 @@ static void GH(tcache_print)(RzCore *core, GH(RTcache) * tcache) {
-static void GH(print_tcache_instance)(RzCore *core, GHT m_arena, MallocState *main_arena, bool main_thread_only) {
-	rz_return_if_fail(core && core->dbg && core->dbg->maps);
+/**
+ * @brief Get a list of RzTcache objects which contain information about tcache.
```
Same here

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1025,278 +1083,656 @@ static void GH(tcache_print)(RzCore *core, GH(RTcache) * tcache) {
+}
+
+/**
+ * @brief Get list of chunks of <bin_num> bin from NBINS array of an arena.
```
Same here

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1025,278 +1083,656 @@ static void GH(tcache_print)(RzCore *core, GH(RTcache) * tcache) {
+}
+
+/**
+ * @brief Get a list of MallocState structs for all the arenas
```
Same here

XVilka commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1025,278 +1083,656 @@ static void GH(tcache_print)(RzCore *core, GH(RTcache) * tcache) {
+}
+
+/**
+ * @brief Get a list of all the heap chunks in an arena. The chunks are in form of a struct RzHeapChunkListItem
```
Same here

XVilka commented on librz/debug/dmap.c:
```diff
@@ -360,3 +88,13 @@ RZ_API RzList *rz_debug_map_list_new(void) {
 }
+
+/**
+ * @brief Get RzList* of memory maps for the process currently being debugged
```
Same here

XVilka commented on librz/include/rz_heap_glibc.h:
```diff
@@ -278,6 +276,17 @@ typedef enum rz_heap_bin_type {
+	ut64 addr;
+	char *type;
+	MallocState *arena;
+} RzArenaListItem;
```
Add a new line after that.

XVilka reviewed (changes requested):
Overall looks good, there are some minor changes required though before being ready to merge.

MalhotraPulak committed:
Implement suggested changes-2

MalhotraPulak committed:
remove unreachable code

MalhotraPulak committed:
change documentation style

MalhotraPulak committed:
Remove `static` keyword and commented code

MalhotraPulak committed:
More documentation fixes

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1370,465 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+					newfile = rz_file_temp("memlib");
+					if (newfile) {
+						file = newfile;
+						rz_core_cmdf(core, "wtf %s 0x%" PFMT64x " @ 0x%" PFMT64x " 2> %s",
```
Just use `rz_file_dump()` here instead.

MalhotraPulak replied:
@XVilka In [this commit](https://github.com/rizinorg/rizin/pull/1182/commits/59691f818c7edd0ba4b6e7726c7ac8ec49b26fe7) commit I used `rz_core_dump()`. I was not able to come up with a testcase which reaches this codeblock so can you just test this once.

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1370,465 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
-		} else
-			cmd_debug_modules(core, input[1]);
+	case '\0': // "dmi" alias of "dmm"
+		rz_core_cmd(core, "dmm", 0);
```
Please use API here too, if possible.

XVilka commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1370,465 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+		return RZ_CMD_STATUS_ERROR;
+	}
+	if (!strcmp(argv[0] + 3, ".*")) {
+		cmd_debug_modules(core, ':');
```
Could you also please tear `cmd_debug_modules` function apart and use the API instead of passing characters/string further?

XVilka replied:
I can see that is uses `int mode` and matching characters instead. I suggest to extract `?` to drop the help, it should be provided by the newshell anyway, and use `RzOutputMode mode` directly.

MalhotraPulak committed:
Use API instead of cmd

MalhotraPulak committed:
Refactor `dmm` and `dmm.` command implementation

XVilka commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -121,8 +121,11 @@ commands:
+              - RZ_OUTPUT_MODE_RIZIN
             summary: List memory map of current module
-            args: []
+            args: [ ]
```
Unnecessary space

MalhotraPulak committed:
minor changes

MalhotraPulak committed:
Made Heap API naming more consistent

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,255 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+}
+
+/* Write the memory map header describing the line columns */
+static void print_debug_map_line_header(RzDebug *dbg) {
```
Maybe remove this function since it's not implemented?

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,255 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+			: rz_str_newf("%08" PFMT64x ".%s", map->addr, rz_str_rwx_i(map->perm));
+		rz_name_filter(name, 0, true);
+		rz_num_units(humansz, sizeof(humansz), map->addr_end - map->addr);
+		dbg->cb_printf("0x%016" PFMT64x " - 0x%016" PFMT64x " %6s %5s %s\n",
```
Since this is located in `librz/core` there is no need to use `cb->printf` here, you could use RzCore and RzCons here freely (this is why we asked to move printing in here after all)/.

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,255 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+					? rz_str_newf("%s.%s", map->name, rz_str_rwx_i(map->perm))
+					: rz_str_newf("%08" PFMT64x ".%s", map->addr, rz_str_rwx_i(map->perm));
+				rz_name_filter(name, 0, true);
+				dbg->cb_printf("f map.%s 0x%08" PFMT64x " 0x%08" PFMT64x "\n",
```
Same here

XVilka commented on librz/core/cdebug.c:
```diff
@@ -639,3 +639,255 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
+			fmtstr = dbg->bits & RZ_SYS_BITS_64 // Prefix formatting string (before bar)
+				? "map %4.8s %c %s0x%016" PFMT64x "%s |"
+				: "map %4.8s %c %s0x%08" PFMT64x "%s |";
+			dbg->cb_printf(fmtstr, humansz,
```
And here

MalhotraPulak committed:
`dbg->cb_printf` -> `rz_cons_printf`

XVilka reviewed (approved):
LGTM, except if there is a quick API to substitute that `rz-bin` spawn. Lets see what @ret2libc and @yossizap think.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,186 @@ commands:
+    summary: Memory map commands
+    subcommands:
+      - name: dm
+        cname: cmd_debug_list_maps
```
use `type` `RZ_CMD_DESC_TYPE_ARGV_STATE` if possible.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,186 @@ commands:
+        cname: cmd_debug_allocate_maps
+        summary: >
+          Allocate <size> bytes at current offset
+          (anywhere if offset is -1)
```
is this really true? It seems to me `addr` is passed to the `map_alloc` method of the dbg plugin, which can do whatever it wants. For debug_native plugin, addr is passed as first argument of `mmap`, which needs NULL to allocate "anywhere", not -1.

MalhotraPulak replied:
Oh, I am not sure about the exact working of this command, this help message was present before the refactor so I just used the same message. Should I remove the ` (anywhere if offset is -1)` part?

ret2libc replied:
For now, yes.

ret2libc commented on librz/core/cmd_descs/cmd_debug.yaml:
```diff
@@ -19,63 +19,186 @@ commands:
+          - name: dmm
+            summary: List modules (libraries, binaries loaded in memory)
+            cname: cmd_debug_modules
+            modes:
```
same, use ARGV_STATE if not too much work. Otherwise, it's ok.

ret2libc commented on librz/core/linux_heap_glibc.c:
```diff
@@ -1406,632 +1836,473 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
+	return RZ_CMD_STATUS_OK;
 }
+
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```
```
[0x55a48dde7d80]> dmhd fastbin 
Tcache in Main Arena @  0x7f2f1eb42a00


Fast bins in Arena @ 0x7f2f1eb42a00
Fast_bin[01] [size: 0x20] Empty bin
Fast_bin[02] [size: 0x30] Empty bin
Fast_bin[03] [size: 0x40] Empty bin
Fast_bin[04] [size: 0x50] Empty bin
Fast_bin[05] [size: 0x60] Empty bin
Fast_bin[06] [size: 0x70] Empty bin
Fast_bin[07] [size: 0x80] Empty bin

Unsorted bin in Arena @ 0x7f2f1eb42a00
Found 0 chunks in unsorted bin

Small bins in Arena @ 0x7f2f1eb42a00
Found 0 chunks in 0 small bins

Large bins in Arena @ 0x7f2f1eb42a00
Found 0 chunks in 0 large bins
```
I would expect only fastbins. Works with other bin_type, but not with fastbins.

MalhotraPulak replied:
Currently `dmhd fast` will show fastbins instead of `dmhd fastbin`, I can add support for both if you want

ret2libc replied:
I think you used `fastbin` in the CHOICES of the argument. Either accept `fastbin` correctly in this command or change the CHOICES to use `fast` instead of `fastbin`.

MalhotraPulak replied:
Yeah my bad, it should have been "fast", anyways I will change it so that it accepts both

XVilka replied:
I think it's better to change CHOICES to `fast` and do not support `fastbin` to be consistent.

MalhotraPulak committed:
use `ARGV_STATE` in `dm`

MalhotraPulak committed:
use `ARGV_STATE` in `dmm`

MalhotraPulak committed:
remove `fastbin` option for dmhd for consistency

MalhotraPulak committed:
use `ARGV_STATE` in `dmh`

XVilka reviewed (approved):
Is this still "WIP"? Looks ready to me.

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1257,84 +1211,75 @@ static int dump_maps(RzCore *core, int perm, const char *filename) {
+	RzListIter *iter;
+	PJ *pj = state->d.pj;
+	RzOutputMode mode = state->mode;
+	if (mode == RZ_OUTPUT_MODE_JSON) {
 		if (!pj) {
 			return;
 		}
 		pj_a(pj);
 	}
```
```suggestion
rz_cmd_state_output_array_start(state);
```

MalhotraPulak replied:
If I remove `pj_a(pj)` then JSON objects will not be inside a list, making it an invalid JSON

ret2libc replied:
`rz_cmd_state_output_array_start` is doing that.

MalhotraPulak replied:
thanks! I missed that

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1257,84 +1211,75 @@ static int dump_maps(RzCore *core, int perm, const char *filename) {
 	}
-beach:
-	if (mode == 'j') {
+	if (mode == RZ_OUTPUT_MODE_JSON) {
 		pj_end(pj);
```
```suggestion
rz_cmd_state_output_array_end(state);
```

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1357,462 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
 }
 
-static int cmd_debug_map(RzCore *core, const char *input) {
+// dm
```
I'm tempted to remove these comments... They will likely become wrong at some point. I'd prefer to focus on https://github.com/rizinorg/rizin/issues/691 to correctly find the right handler.

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1357,462 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
 		} else {
-			eprintf("Usage: dmL addr size\n");
-			return false;
+			return RZ_CMD_STATUS_WRONG_ARGS;
```
```suggestion
			return RZ_CMD_STATUS_ERROR;
```
WRONG_ARGS usually means you passed the wrong number of args, but in this case it is something else.

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1357,462 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+	rz_debug_map_sync(core->dbg); // update process memory maps
+	RzList *list = rz_debug_modules_list(core->dbg);
+	rz_list_foreach (list, iter, map) {
+		if ((!libname ||
```
very very long `if`. Not sure if it comes from old code, but a separate function with a meaningful name or something else would be better

MalhotraPulak replied:
Yeah this `if` comes from the old code indeed

ret2libc commented on librz/core/cmd_debug.c:
```diff
@@ -1443,376 +1357,462 @@ static bool get_bin_info(RzCore *core, const char *file, ut64 baseaddr, PJ *pj,
+// dmx
+RZ_IPI int rz_cmd_debug_heap_jemalloc(void *data, const char *input) {
+	RzCore *core = (RzCore *)data;
+	if (rz_debug_is_dead(core->dbg)) {
+		rz_cons_println("Debugging is not enabled. Run ood?");
+		return RZ_CMD_STATUS_ERROR;
+	}
```
```suggestion
#define CMD_CHECK_DEBUG_DEAD(core) do { \
  if (rz_debug_is_dead(core->dbg)) { \
    rz_cons_println("Debugging is not enabled. Run ood?"); \
    return RZ_CMD_STATUS_ERROR; \
  }
} while(0)

	CMD_CHECK_DEBUD_DEAD(core);
```
better to avoid all the repetitions. This is probably the easy thing to reduce that.

ret2libc commented:
Overall looks good anyway! Great work ;)

MalhotraPulak committed:
Replaced repetitive code with macro

MalhotraPulak committed:
Use `rz_cmd_state_output_array`

XVilka committed:
Merge branch 'dev' into refactor-dm

ret2libc reviewed (approved):
It looks good to me, but since you are touching a lot of debug code (which is not tested too much) please do a round of manual testing with a debugging session. I did it already a couple of versions ago, but better be sure ;)

yossizap commented on librz/include/rz_core.h:
```diff
@@ -772,6 +773,22 @@ RZ_API char *rz_core_sysenv_begin(RzCore *core, const char *cmd);
+RZ_API RzList *rz_heap_heap_chunks_list_32(RzCore *core, MallocState *main_arena, ut32 m_arena, ut32 m_state);
+RZ_API bool rz_heap_resolve_main_arena_32(RzCore *core, ut32 *m_arena);
+RZ_API bool rz_heap_update_main_arena_32(RzCore *core, ut32 m_arena, MallocState *main_arena);
+RZ_API RzList *rz_heap_tcache_list_32(RzCore *core, ut32 m_arena, MallocState *main_arena, bool main_thread_only);
```
Since we are planning to use the API here for Cutter's HeapWidget frontend the _32/_64 variations in the API should be made internal and replaced with wrapper functions that will use a bool/bitness var to differentiate architectures.

XVilka replied:
@PulakIIIT I think this can be done in a separate PR. Let's merge this one.

yossizap reviewed (approved):
I am a bit late for this review so I'll just leave the comment regarding the API for a future PR. LGTM other than that.
